source/agora/common/Deserializer.d

```diff
@@ -149,61 +149,63 @@ public T deserializeFull (T) (scope DeserializeDg dg) @safe
 }
 
 /// Ditto
-public void deserializePart (T) (ref T record, scope DeserializeDg dg) @safe
-    if (isInputRange!T && hasLength!T)
-{
-    size_t length = deserializeVarInt!size_t(dg);
-    record.length = length;
-    foreach (ref v; record)
-        deserializePart(v, dg);
-}
-
-/// Ditto
-public void deserializePart (T) (ref T record, scope DeserializeDg dg) @safe
-    if (is(T == struct))
+public void deserializePart (T) (
+    ref T record, scope DeserializeDg dg,
+    CompactMode compact = CompactMode.Yes)
+    @safe
 {
     import geod24.bitblob;
 
+    // Custom deserialization trumps everything
     static if (hasDeserializeMethod!T)
         record.deserialize(dg);
+
     // BitBlob are fixed size and thus, value types
     // If we use an `ubyte[]` overload, the deserializer looks for the length
     else static if (is(T : BitBlob!N, size_t N))
         record = T(dg(T.Width));
-    else
-        foreach (const ref field; record.tupleof)
-            deserializePart(field, dg);
-}
 
-/// Enum support
-public void deserializePart (T)(ref T record, scope DeserializeDg dg)
-    @trusted
-    if (is(T == enum))
-{
-    OriginalType!T orig_val;
-    deserializePart(orig_val, dg);
-    record = cast(T)(orig_val);
-}
+    // Range deserialization
+    else static if (isInputRange!T && hasLength!T)
+    {
+        size_t length = deserializeVarInt!size_t(dg);
+        record.length = length;
+        foreach (ref v; record)
+            deserializePart(v, dg);
+    }
 
-/// Ditto
-public void deserializePart (T) (ref T record, scope DeserializeDg dg,
-    CompactMode compact = CompactMode.Yes)
-    @trusted
-    if (isScalarType!T)
-{
-    static if (is(Unqual!T == bool))
+    // Enum deserialize as their base type
+    else static if (is(T == enum))
+    {
+        OriginalType!T orig_val;
+        deserializePart(orig_val, dg);
+        record = cast(T)(orig_val);
```

Comment:
What about `assert(orig_val >= T.min && orig_val <= T.max);`? Although I'm not sure if this works well with things like `enum E { f = "foo", b = "bar" }`..

Reply:
I didn't wanna make any assumption about the type of `enum`. I don't think we support enum strings though, only enum based on integer. There are quite a few places where we need to sanitize the input: most notably, length of array (someone could DoS us with just a bit of binary data...). Maybe we should open an issue for this ?